Modules/_posixsubprocess.c

```diff
@@ -453,11 +453,13 @@ child_exec(char *const exec_array[],
 
     /* Close pipe fds.  Make sure we don't close the same fd more than */
     /* once, or standard fds. */
-    if (p2cread > 2)
+    if (p2cread > 2 && !_is_fd_in_sorted_fd_sequence(p2cread, py_fds_to_keep))
```

Comment:
Note that there is no need to manually close `p2cread` at all, and other two descriptors need to be closed only if they are the result of `dup` above. This is because if a descriptor is created by subprocess.py (a pipe or for `/dev/null`), it's non-inheritable, and if a descriptor is specified by the user, is not in `pass_fds` and is > 2, it will be closed by `_close_open_fds`. So my original idea of fixing this issue was to add an async-signal-safe version of `_Py_dup` and use it instead of `dup` above, and then just remove these three checks.

Note also that `p2cwrite` and its three friends are never required to be manually closed since they are always non-inheritable pipe ends created by subprocess.py.

Reply:
Not surprising. The original code here predates our non-inheritable file descriptor implementation. I'm examining this vs your https://github.com/izbyshev/cpython/commit/b89b52f28490b69142d5c061604b3a3989cec66c changes to see what makes the most sense.

This change as is _does_ make it pass your unittest.